File: backend/src/main/java/org/example/echo/service/ArtistService.java
package org.example.echo.service;

import lombok.RequiredArgsConstructor;
import org.example.echo.dto.album.AlbumResponse;
import org.example.echo.dto.artist.ArtistCreateRequest;
import org.example.echo.dto.artist.ArtistResponse;
import org.example.echo.dto.artist.ArtistUpdateRequest;
import org.example.echo.dto.track.TrackResponse;
import org.example.echo.entity.Artist;
import org.example.echo.entity.Track;
import org.example.echo.exception.custom.EntityAlreadyExistsException;
import org.example.echo.exception.custom.EntityNotFoundException;
import org.example.echo.mapper.AlbumMapper;
import org.example.echo.mapper.ArtistMapper;
import org.example.echo.mapper.TrackMapper;
import org.example.echo.repository.AlbumRepository;
import org.example.echo.repository.ArtistRepository;
import org.example.echo.repository.TrackRepository;
import org.example.echo.service.storage.StorageService;
import org.springframework.core.io.Resource;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ArtistService {
    private final ArtistRepository artistRepository;
    private final TrackRepository trackRepository;
    private final AlbumRepository albumRepository;
    private final StorageService storageService;

    public ArtistResponse createArtist(ArtistCreateRequest request) {
        if (artistRepository.existsByName(request.name()))
            throw new EntityAlreadyExistsException("Artist", request.name());

        Artist artist = Artist
                .builder()
                .name(request.name())
                .build();

        return ArtistMapper.toResponse(artistRepository.save(artist));
    }

    public Page<ArtistResponse> getAllArtists(Pageable pageable) {
        return artistRepository.findAll(pageable)
                .map(ArtistMapper::toResponse);
    }

    public ArtistResponse getArtistById(Long id) {
        Artist artist = artistRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Artist", id));

        return ArtistMapper.toResponse(artist);
    }

    public Page<TrackResponse> getArtistTracks(Long id, Pageable pageable) {
        if (!artistRepository.existsById(id)) {
            throw new EntityNotFoundException("Artist", id);
        }

        return trackRepository.findByArtistId(id, pageable)
                .map(TrackMapper::toResponse);
    }

    public Page<AlbumResponse> getArtistAlbums(Long id, Pageable pageable) {
        if (!artistRepository.existsById(id)) {
            throw new EntityNotFoundException("Artist", id);
        }

        return albumRepository.findByArtistId(id, pageable)
                .map(AlbumMapper::toResponse);
    }

    public List<ArtistResponse> searchArtistByName(String q) {
        return artistRepository.searchByName(q).stream().map(ArtistMapper::toResponse).toList();
    }

    public Resource downloadArtistAsZip(Long id) {
        Artist artist = artistRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Artist", id));

        List<String> filePaths = trackRepository.findByArtistId(id)
                .stream()
                .map(Track::getAudioUrl)
                .toList();

        return storageService.loadFilesAsZip(filePaths, artist.getName().replaceAll("\\s+", "_"));
    }

    @Transactional
    public ArtistResponse updateArtist(Long id, ArtistUpdateRequest request) {
        Artist artist = artistRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Artist", id));

        if (artistRepository.existsByName(request.name()))
            throw new EntityAlreadyExistsException("Artist", request.name());

        artist.setName(request.name());

        return ArtistMapper.toResponse(artistRepository.save(artist));
    }

    @Transactional
    public void deleteArtist(Long id) {
        Artist artist = artistRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Artist", id));

        artistRepository.delete(artist);
    }
}

File: frontend/src/api/artistService.ts
import apiClient from './apiClient';
import { ArtistResponse, ArtistCreateRequest, ArtistUpdateRequest, PageResponse, AlbumResponse, TrackResponse } from '../types';

export const artistService = {
  async getAllArtists(page: number = 0, size: number = 10): Promise<PageResponse<ArtistResponse>> {
    const response = await apiClient.get<PageResponse<ArtistResponse>>('/artists', {
      params: { page, size },
    });
    return response.data;
  },

  async getArtistById(id: number): Promise<ArtistResponse> {
    const response = await apiClient.get<ArtistResponse>(`/artists/${id}`);
    return response.data;
  },

  async searchArtists(query: string): Promise<ArtistResponse[]> {
    const response = await apiClient.get<ArtistResponse[]>('/artists/search', {
      params: { q: query },
    });
    return response.data;
  },

  async createArtist(data: ArtistCreateRequest): Promise<ArtistResponse> {
    const response = await apiClient.post<ArtistResponse>('/artists', data);
    return response.data;
  },

  async updateArtist(id: number, data: ArtistUpdateRequest): Promise<ArtistResponse> {
    const response = await apiClient.put<ArtistResponse>(`/artists/${id}`, data);
    return response.data;
  },

  async deleteArtist(id: number): Promise<void> {
    await apiClient.delete(`/artists/${id}`);
  },

  async getArtistAlbums(id: number, page: number = 0, size: number = 10): Promise<PageResponse<AlbumResponse>> {
    const response = await apiClient.get<PageResponse<AlbumResponse>>(`/artists/${id}/albums`, {
      params: { page, size },
    });
    return response.data;
  },

  async getArtistTracks(id: number, page: number = 0, size: number = 10): Promise<PageResponse<TrackResponse>> {
    const response = await apiClient.get<PageResponse<TrackResponse>>(`/artists/${id}/tracks`, {
      params: { page, size },
    });
    return response.data;
  },

  async downloadArtistTracks(id: number): Promise<Blob> {
    const response = await apiClient.get(`/artists/${id}/download`, {
      responseType: 'blob',
    });
    return response.data;
  },
};

